### ISSUE TITLE
IncuCyte HCS Example Data

### ISSUE THREAD

**Issue by constantinpape** [mobie/mobie-viewer-fiji#971]
I have added example data with two wells. This data has only one channel, but it contains multiple timepoints per position (here: 4).

The naming pattern is as follows:
`<PREFIX>_<WELL_NAME>_<POSITION>_<TIMEPOINT>` (e.g. `MiaPaCa2-PhaseOriginal_A2_1_03d06h40m.tif`)

The example data is here: https://owncloud.gwdg.de/index.php/s/HfabNGN7tAFw81s

**Comment by tischi** [comment]
Do you know how that _would_ look like if it had more than one channel?

**Comment by tischi** [comment]
@constantinpape i

If you have a clever idea how we could represent multiple files for multiple time points in our `StorageLocation`, let me know :-)

```
// create site image source
final StorageLocation storageLocation = new StorageLocation();
storageLocation.absolutePath = hcsPlate.getPath( channel, well, site );
storageLocation.channel = 0;

final ImageDataSource imageDataSource = new ImageDataSource( hcsPlate.getSiteKey( channel, well, site ), ImageDataFormat.ImageJ, storageLocation );
dataset.addDataSource( imageDataSource );
```

Maybe we could support a regex pattern as `storageLocation.absolutePath` in mobie-io?

**Comment by tischi** [comment]
Maybe better:

```
Map< Integer, Map< Integer, String > >  storageLocation.absolutePathsZT;
```

**Comment by constantinpape** [comment]
> If you have a clever idea how we could represent multiple files for multiple time points in our `StorageLocation`, let me know :-)

Hmm, sorry I am not quite sure what exactly the issue is - I think we would need to discuss this via phone / zoom for me to understand.

> Do you know how that would look like if it had more than one channel?

Yes, the channels are stored in different sub-folders (each following the same naming pattern as described above). I have uploaded another example with two channels (phase-contrast as before and an additional red fluorescence channel) here:
https://owncloud.gwdg.de/index.php/s/n8ueCIHTLaLrNGJ

**Comment by tischi** [comment]
> Hmm, sorry I am not quite sure what exactly the issue is - I think we would need to discuss this via phone / zoom for me to understand.

For us, one `Source` is a XYZT image and the `StorageLocation` contains the information where to load the corresponding XYZT data. We never had the use-case where this XYZT data is distributed across multiple files, thus we cannot represent this in our [current implementation of StorageLocation](https://github.com/mobie/mobie-viewer-fiji/blob/main/src/main/java/org/embl/mobie/lib/io/StorageLocation.java). 

Now, to load HCS data like yours where multiple time-points are distributed across multiple files we need something in `StorageLocation` that can point to multiple files (and express that those files correspond to time-points (and not z-planes)),

**Comment by tischi** [comment]
@constantinpape I uploaded something for you to test on MoBIE-beta.

New menu command `Open>Open HCS Dataset...`

**Comment by constantinpape** [comment]
Thanks @tischi! 

I can open data from an example screen (bigger than what I send you), but there are a few issues:

1. The well boundaries are misaligned (see the first screenshot below).

2. The name overlay doesn't print the proper well names, but just `ch_1` (see also first screenshot).

3. The time ordering is wrong and seems to be random, see the second image, which shows the first three consecutive frames for one position in MoBIE, and the actual ones.

![image](https://user-images.githubusercontent.com/4263537/221789753-1ab69b80-3a5f-4785-9334-0f370611c650.png)

![mobie-time](https://user-images.githubusercontent.com/4263537/221794861-68c880f4-f4ea-4f66-81a7-134a2b0170b4.png)

**Comment by tischi** [comment]
1. For the initial test data that I got from you they are well aligned, could you share example data with me to reproduce this?
2. Did this ever work (i.e. in covid-if) in terms of showing well names? I don't think so...
3. I fixed it.

**Comment by constantinpape** [comment]
> 1. For the initial test data that I got from you they are well aligned, could you share example data with me to reproduce this?

I send you the location on the EMBL server in mattermost.

> 2. Did this ever work (i.e. in covid-if) in terms of showing well names? I don't think so...

Yes you're right. (I thought about a different example with a transformed grid; where I think it works). It would be great to get it to work for the merged grid, since seeing the well names is quite helpful. But I understand that this might be a bit complex due to the hierarchy of sources...

> 3\. I fixed it.

:+1:

**Comment by tischi** [comment]
A likely cause of issue is that there is only one site per well. Probably there is a bug when creating a MergedGrid with only one image :-) 

1.  fix the bug for creating a MergedGrid with only one image.
2. change to code not to create a MergedGrid if there is only one image.

**Comment by tischi** [comment]
OK, you really have found an edge case here :-)

It is only one site, but this site is not at `[0,0]` within the well but at `[0,1]`, because it is site #2.

I could not find the bug yet.

So step 1, I changed the code such that if there is only one site per well that it puts it at grid position `[0,0]` no matter the site name.

This works:

<img width="986" alt="image" src="https://user-images.githubusercontent.com/2157566/221887723-f12669a8-2723-45c2-b26c-50bab05e92f4.png">

Next step: 

- [ ] Make it not use a StitchedImage at all if there is only one site.

**Comment by tischi** [comment]
@constantinpape I uploaded a version that should work for you.

**Comment by tischi** [comment]
If it works, could we close this issue in favour of more details requests such as well name overlays?

**Comment by constantinpape** [comment]
Yep, it works now! I will open a separate issue about the well name overlays for HCS data.
